src/utils/DockerInstallStatusProvider.ts

```diff
@@ -0,0 +1,33 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AsyncLazy } from "./lazy";
+import { execAsync } from "./spawnAsync";
+
+export interface IDockerInstallStatusProvider {
+    isDockerInstalled(): Promise<boolean>
+}
+
+class DockerInstallStatusProvider implements IDockerInstallStatusProvider {
+    private maxCacheTime: number = 5000;
+    private isDockerInstalledLazy: AsyncLazy<boolean>;
+
+    public constructor() {
+        this.isDockerInstalledLazy = new AsyncLazy<boolean>(async () => {
+            try {
+                await execAsync('docker -v');
+                return true; // As long as the docker command did't throw exception, assume it is installed.
```

Comment:
@ravipal I do have another suggestion; once Docker is installed it's _very_ unlikely that it will become uninstalled during the session. Once we get installed = true, we can skip the `docker -v` calls from then on.

Worst case, if it does get uninstalled, then we'll show the "not running" error instead of the "not installed" error, which I'm totally OK with given the infrequency of the problem.

Reply:
This would be beneficial in the probably much-more-common case that Docker is installed but not running, by avoiding repeated `docker -v` calls.

Reply:
If docker is installed, then this method wouldn't get called. For docker not running case, we still should continue to check the status.

Reply:
Ah, I see that is implemented via `this._currentDockerStatus` in `LocalRootTreeItemBase`. Still, I would consider moving this functionality into this class; it would make `LocalRootTreeItemBase` a little cleaner and abstract away install status, leaving it here where it fits better IMO.

Reply:
I like @bwateratmsft idea. If we determine during a session that Docker is installed, subsequent failure to refresh any Docker pane (containers, images, etc) is enough to consider Docker as "not running". We do not need to repeat `docker -v` check.

Reply:
If done that way we can remove `DockerStatus` / `_currentDockerStatus`, and let the refresh logic be:

1. No error? Yay!
1. Error. Is Docker installed?
    1. Yes: not running
    1. No: not installed

Reply:
An additional thought; installing Docker doesn't (currently) cause it to be automatically started, and I think it would be a great user experience if the UI magically transitioned from "Docker isn't installed" to "Docker isn't running".

Reply:
I am not sure I follow the suggestion. `DockerInstallStatusProvider` provide the current docker install status either installed or not installed and it caches the value for 5 seconds to optimize for perf. The caller has a timer that periodically tries to refresh the node and if it fails to refresh, then checks for docker status and if it is changed, then it decides to refresh the tree view.
The current implementation will update the view from 'Docker not installed' to 'Docker not running' id docker is installed.

Reply:
If it only gets called once, then the UI can't change without restarting VSCode.

Reply:
Now docker status is cached forever once it is installed.